### PR TITLE
chore: ScheduleInternal refactoring

### PR DESCRIPTION
**Pull request by romange**
A small refactoring to improve the flow of ScheduleInternal() as well as to prepare it for the next change that will reduce the CPU load from the shard queue.

### CODE REVIEW

File: src/server/transaction.cc [hunk 1]
```diff
@@ -703,35 +712,39 @@ void Transaction::ScheduleInternal() {
...
-    atomic_uint32_t schedule_fails = 0;
-    auto cb = [this, &schedule_fails, can_run_immediately]() {
-      if (!ScheduleInShard(EngineShard::tlocal(), can_run_immediately)) {
-        schedule_fails.fetch_add(1, memory_order_relaxed);
-      }
-      FinishHop();
-    };
-
     run_barrier_.Start(unique_shard_cnt_);
+
```
Comment (romange) [inline src/server/transaction.cc:716]: there are no functional changes in this code

File: src/server/transaction.cc [hunk 2]
```diff
@@ -703,35 +712,39 @@ void Transaction::ScheduleInternal() {
...
-      // during tests.
-      ThisFiber::PrintLocalsCallback locals([&] {
-        return absl::StrCat("unique_shard_cnt_: ", unique_shard_cnt_,
-                            " run_barrier_cnt: ", run_barrier_.DEBUG_Count(), "\n");
-      });
-      run_barrier_.Wait();
+      break;
     }
 
-    if (schedule_fails.load(memory_order_relaxed) == 0) {
```
Comment (romange) [inline src/server/transaction.cc:731]: if we go into CanRunInlined branch then schedule_fails is 0. In fact schedule_fails check is only relevant for the else branch (IterateActiveShards)

File: src/server/transaction.cc [hunk 3]
```diff
@@ -762,6 +775,9 @@ void Transaction::ScheduleInternal() {
       });
     }
   }
+
+  coordinator_state_ |= COORD_SCHED;
```
Comment (romange) [inline src/server/transaction.cc:779]: final phase: we succeeded scheduling - update the stats

File: src/server/transaction.cc (hunk 2 above)
Comment (dranikpg) [reply src/server/transaction.cc:731]: True, good simplification

File: src/server/transaction.cc [hunk 4]
```diff
@@ -703,35 +712,39 @@ void Transaction::ScheduleInternal() {
...
-      coordinator_state_ |= COORD_SCHED;
+    ScheduleContext schedule_ctx{this, can_run_immediately};
+
+    auto cb = [&schedule_ctx]() {
+      if (!schedule_ctx.trans->ScheduleInShard(EngineShard::tlocal(),
+                                               schedule_ctx.optimistic_execution)) {
+        schedule_ctx.fail_cnt.fetch_add(1, memory_order_relaxed);
+      }
+      schedule_ctx.trans->FinishHop();
+    };
```
Comment (dranikpg) [inline src/server/transaction.cc:735]: 1. You forgot a rename in the last pr: can_run_immediately 🙂 
2. What's the raison d'être of ScheduleContext? 🤓 You don't pass it around, so it's not code simplicifaction. Is it to reduce the functor size by capturing a single pointer?

**Review by dranikpg — APPROVED** [review]

File: src/server/transaction.cc (hunk 4 above)
Comment (romange) [reply src/server/transaction.cc:735]: 1. Will do.
2. You will see it in the next PR (if it works)

**Pull request merged**